File: 331-验证二叉树的前序序列化.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        vector<int> stack(1, 1);
        int i = 0;

        while(i < preorder.size()){
            if(preorder[i] != ','){
                if(preorder[i] == '#'){
                    if(stack.size() > 0 && stack[stack.size() - 1] > 0){
                        stack[stack.size() - 1]--;
                        i++;
                        if(stack[stack.size() - 1] == 0) stack.pop_back();
                    }
                    else{
                        return false;
                    }
                }
                else{
                    while(i < preorder.size() && preorder[i] != ',') i++;
                    if(stack.size() > 0 && stack[stack.size() - 1] > 0){
                        stack[stack.size() - 1]--;
                        if(stack[stack.size() - 1] == 0) stack.pop_back();
                        stack.push_back(2);
                    }
                    else{
                        return false;
                    }
                }
            }
            else{
                i++;
            }
        }

        return stack.size() == 0;
    }
};

File: 331-验证二叉树的前序序列化.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        preorder = preorder.split(',')
        index = self.helper(preorder, 0)

        return True if index == len(preorder) else False

    def helper(self, preorder, index):
        if index >= len(preorder):
            return -1
        if preorder[index] == '#':
            return index + 1
        index += 1
        index = self.helper(preorder, index)
        if index == -1:
            return -1
        index = self.helper(preorder, index)
        if index == -1:
            return -1

        return index
